crates/typst-realize/src/lib.rs

```diff
@@ -670,6 +671,21 @@ fn visit_grouping_rules<'a>(
         }
 
         finish_innermost_grouping(s)?;
+        i += 1;
+        if i > 4096 {
```

Comment:
Maybe the limit is too small for complex documents?

Reply:
In which sane cases would it be hit?

Reply:
I couldn't really come up with one, so I guess it *should* be fine

Reply:
So I've reduced it to 512, because I couldn't really find any real world cases where this would go above like 3, and otherwise the stack will overflow in a test.

Reply:
is this, in sane cases, upper bounded by the number of different grouping rules essentially? hence 3?

Reply:
So I did some more digging, and I've found this test case that goes up to `24`, and increases the longer the text and words get:
```typ
// --- show-text-regex-character-class ---
// This is a fun one.
#set par(justify: true)
#show regex("\S"): letter => box(stroke: 1pt, inset: 2pt, upper(letter))
#lorem(5)

// trailing blank line is important!
```

Reply:
But I guess these cases are so slow and consume so much memory, that it's sensible to have an upper bound for them.